### ISSUE TITLE
`go.Mod.UpstreamProviderOrg` is unreliable and misleading

### ISSUE THREAD

**Issue by guineveresaenger** [pulumi/upgrade-provider#213]
### What happened?

For the new [upstream upgrade check](https://github.com/pulumi/pulumi-gitlab/blob/master/.github/workflows/check-upstream-upgrade.yml) we use `go.Mod.UpstreamProviderOrg` to detect upstream releases.

However, the way that field gets populated is by attempting to read the upstream provider from the `provider/go.mod` file, which:

- returns the first found match for `*/terraform-provider-foo` 
- does not account for submodules that have since been reamed (such as [DataDog](https://github.com/pulumi/pulumi-datadog/blob/master/provider/go.mod#L17) - the org is `DataDog` not `terraform-providers`.
- does not account for gitlab having a [mirror repository](https://github.com/gitlabhq/terraform-provider-gitlab), which [never makes it into our go.mod](https://github.com/pulumi/pulumi-gitlab/blob/master/provider/go.mod#L18)

- in summary, may or may not be completely different from the relevant upstream org/repo that we would want to check versions from.

This behavior is correct for "typical" providers with no renames, replaces, or otherwise obscuring the link to the GitHub org/repo location of the upstream provider, so it "works" in those instances.

The reason we did not detect this bug sooner is that we did not ever actually use the `go.Mod.UpstreamProviderOrg` before attempting to discover upstream upgrades.

### Example

https://github.com/pulumi/pulumi-gitlab/issues/445

### Output of `pulumi about`

n/a

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).
